### ISSUE TITLE
Rename `Node.remove_child()` to `Node.detach_child()`

### ISSUE THREAD

**Issue by SilencedPerson** [godotengine/godot-proposals#4207]
### Describe the project you are working on

Teaching teenagers how to use and program with Godot.

### Describe the problem or limitation you are having in your project

I frequently run into problem where people understand `remove_child()` as `delete_child()`. This is very bad as it leads to memory leaks.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The used language should be more direct. Word "remove" seems to imply deletion.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I am suggesting `detach_child()` but I would love to hear your takes.
One problem I have with it is that it kind of begs to rename `add_child()` to `attach_child()` to introduce parity but that method doesn't seem to cause any problems.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Removing nodes is frequent task in Godot. I don't think there is easy way to going about it.

### Is there a reason why this should be core and not an add-on in the asset library?

Cannot be done with add-on.

**Comment by Mickeon** [comment]
I do agree on a rename to an extent. Although, I'm not sure about "detach" itself. It does make sense, but...

> Removing nodes is frequent task in Godot. I don't think there is easy way to going about it.

Indeed, "Removing". It feels like just the right word for it, like it is for Arrays. The memory leak is more of an issue of implementation of the Nodes themselves.

You could consider bringing this suggestion to the attention of this tracker:
https://github.com/godotengine/godot/issues/54161#issuecomment-1063503415

**Comment by zinnschlag** [comment]
Related to #4156 ? This one is about C#. But the same could be applied to GDScript.

**Comment by caspervonb** [comment]
Granted naming things is hard but there's plenty of prior art for tree structures with add_child/remove_child.
IMHO this would be step towards worse naming.

**Comment by Mickeon** [comment]
I think the proposal strengthens the notion that Nodes outside the tree being inherently prone to memory leaks are problematic. Also hints that, because Nodes are so important in Godot, the documentation for `free()` and `queue_free()` should mention what happens before a Node is freed, it being they're safely removed from the SceneTree.

**Comment by diddykonga** [comment]
I can partially understand the confusion but it really relies upon your prior understanding about how Nodes work which is a core part of the engine.
Also, for `add_child` it is more easy to tell what it does by looking at the parameters and lack of return, which allows you to infer that it does not create anything for you, which could then lead one to understand that `remove_child` does not delete anything for you.

**Comment by Mickeon** [comment]
It's also worth noting that in Godot 4.0, **Array**'s `remove()` (in 3.x called `erase()`) actually removes by value/reference, which in a way is furthermore  similar to **Node**'s `remove_child()`. This way, and with slightly more elaborate documentation on `remove_child()` about `free()` and `queue_free()`, the confusion would be mitigated already.

I wanted to further back this up with the Dictionary function, but it seems like [Dictionary still uses `erase()`](https://docs.godotengine.org/en/latest/classes/class_dictionary.html#class-dictionary-method-erase), which is quite the inconsistency...

**Comment by Mickeon** [comment]
With that said, however, if a rename has to go through. If I may suggest:
`Node.remove_child()` -> `Node.expel_child()` or
`Node.remove_child()` -> `Node.eject_child()`

`Node.expel_child()` being my primary choice. As humorous as they may be to read without context, both names aim to further highlight that the child is not deleted from memory, while (especially in the case of "expel") being synonyms of "remove".

**Comment by mieldepoche** [comment]
Adding a new `remove_and_queue_free_child` or `remove_and_free_child` method would fix ambiguity: one would infer that `remove_child` does indeed not free anything. It would basically be `queue_free` from what I understand so it would be more of a intended redundancy than an actual feature. Still worth it imo.

**Comment by MewPurPur** [comment]
Naaah, we shouldn't have remove_and_free_child() when it has the same function as remove_child(); free(), that's just redundant.

I like the proposal, I've been using Godot for a while now and I'm only learning this now. I don't see why we wouldn't just do `add_child --> attach_child()` and `remove_child() -> detach_child()`. It is more self-explanatory, the English is familiar, and the two words are the exact opposite, so it's easier to understand and learn.

**Comment by KingBrainlord** [comment]
It'd be helpful, perhaps, if a set of antonyms could be used for "add_child" and "remove" child so that its more obvious that they are reverse operations. Maybe "inject_child" and "eject_child"? That sounds bad though. Maybe "insert_child" and "withdraw_child"? "attach" and "detach"? connect and disconnect? Or even "adopt child" and "disown child"? "accept child," and "reject child"? Sorry, some of these are just fun.

**Comment by octetdev2** [comment]
Seems like a strange thing to rename to me. A tree is a very well understood structure, the immediate thing you look for in the API is add/remove/get child operations.

Are we going to invent new terms for array, set and map operations too? ejecting values arrays? 🙃

The fact that nodes are not reference counted in GDScript is something that should be clearly and boldly documented. Perhaps we could even do static analysis on the GDScript side of things and emit warnings when a node is removed and never freed.

**Comment by me2beats** [comment]
add_child() remove_child() looks fine.
removing is like moving away, not deleting.
also remove_child sounds consistent with move_child imo

Attach/detach sounds a bit more complicated
I believe most frequently used API names should be(sound) simple

And let's just count how many engines/frameworks use add/remove and attach/detach.

**Comment by pycbouh** [comment]
Here's a crazy one: `orphan_child` describes the action perfectly and matches nicely with the idea of orphan nodes presented elsewhere in the engine 😛 

However, this would be a very weird naming format for anyone familiar with programming, where add/remove are an expected, widely used pair of verbs for the opposite actions. Making syntax more ELI5 can make it harder to use for more experienced people. Though, every language and framework has its peculiarities, of course...
